bittide-instances/tests/Wishbone/ScatterGather.hs

```diff
@@ -0,0 +1,111 @@
+--
+-- SPDX-License-Identifier: Apache-2.0
+{-# LANGUAGE NumericUnderscores #-}
+{-# OPTIONS_GHC -fplugin=Protocols.Plugin #-}
+
+module Wishbone.ScatterGather where
+
+import Clash.Explicit.Prelude
+import Clash.Prelude (HiddenClockResetEnable, withClockResetEnable)
+
+import Data.Char (chr)
+import Data.Maybe (mapMaybe)
+import Language.Haskell.TH
+import Project.FilePath
+import Protocols
+import Protocols.Idle
+import System.FilePath
+import Test.Tasty
+import Test.Tasty.HUnit
+import Test.Tasty.TH
+import VexRiscv (DumpVcd (NoDumpVcd))
+
+import Bittide.Calendar
+import Bittide.DoubleBufferedRam
+import Bittide.ProcessingElement
+import Bittide.ProcessingElement.Util
+import Bittide.ScatterGather
+import Bittide.SharedTypes
+import Bittide.Wishbone
+
+import qualified Protocols.Df as Df
+import qualified Prelude as P
+
+case_scatter_gather_echo_test :: Assertion
+case_scatter_gather_echo_test = do
+  assertBool
+    msg
+    (P.head (lines simResult) == "Written data was read back correctly")
+ where
+  msg =
+    "Received the following from the CPU over UART:\n"
+      <> unlines (P.take 5 (lines simResult))
+  simResult = chr . fromIntegral <$> mapMaybe Df.dataToMaybe uartStream
+
+  uartStream =
+    sampleC def
+      $ withClockResetEnable clockGen resetGen enableGen
+      $ dut @System @4 @32 scatterConfig gatherConfig
+      <| idleSource
+
+  scatterConfig = ScatterConfig $ CalendarConfig (SNat @32) scatterCal scatterCal
+  gatherConfig = GatherConfig $ CalendarConfig (SNat @32) gatherCal gatherCal
+
+  -- Padding is required to increase the duration of a metacycle, giving the CPU
+  -- enough time to write to the gather memory and read from the scatter memory.
+  -- The calendar for the scatter unit is delayed by one cycle.
+  padding = 1024
+  incrementingCal = genIncrementingCalendar @16
+  gatherCal = incrementingCal :< ValidEntry maxBound (padding - 1 :: Unsigned 16)
+  scatterCal = (ValidEntry 0 0 :> incrementingCal) :< ValidEntry maxBound (padding - 2)
+
+genIncrementingCalendar ::
+  forall size repititionBits.
+  (KnownNat size, KnownNat repititionBits) =>
+  Calendar size (Index size) repititionBits
+genIncrementingCalendar = iterateI f (ValidEntry 0 0)
+ where
+  f (ValidEntry i _) = ValidEntry (succ i) 0
+
+{- | A simulation-only instance containing just VexRisc with UART and a scatter and a
+gather unit. The VexRiscv runs the `scatter_gather_test` binary from `firmware-binaries`.
+-}
+dut ::
+  forall dom nBytes addrW.
+  ( HiddenClockResetEnable dom
+  , KnownNat nBytes
+  , 1 <= nBytes
+  , KnownNat addrW
+  ) =>
+  ScatterConfig nBytes addrW ->
+  GatherConfig nBytes addrW ->
+  Circuit (Df dom (BitVector 8)) (Df dom (BitVector 8))
+dut scatterConfig gatherConfig = circuit $ \uartRx -> do
+  jtagIdle <- idleSource -< ()
+  [uartBus, wbSu, wbGu] <- processingElement NoDumpVcd peConfig -< jtagIdle
+  (uartTx, _uartStatus) <- uartInterfaceWb d2 d2 uartSim -< (uartBus, uartRx)
```

Comment:
You might want to increase the tx buffer size slightly to prevent the CPU from waiting during the prints.
`d8` or `d16` should be fine.